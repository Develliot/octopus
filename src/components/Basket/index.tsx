import { memo } from 'react';
import { Span } from 'src/components/Typography';
import { ProductQuantityType } from 'src/types';

type Props = {
    basket: ProductQuantityType[];
};

export const basketTotal = (basket: ProductQuantityType[]): number =>
    basket.reduce(
        (total, currentProduct) =>
            total + currentProduct.product.price * currentProduct.quantity,
        0
    );

export const Basket = ({ basket }: Props) => {
    return (
        <div>
            {!basket.length && <Span>Empty Basket</Span>}
            {basket.length
                ? basket.map((basketItem: ProductQuantityType) => (
                      <div key={basketItem.product.id}>
                          <Span>
                              {basketItem.product.title} : {basketItem.quantity}
                          </Span>
                      </div>
                  ))
                : null}
            <Span>Total cost: Â£{basketTotal(basket).toFixed(2)}</Span>
        </div>
    );
};

export default memo(Basket);
