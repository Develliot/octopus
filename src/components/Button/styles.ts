import styled from 'styled-components';

export const ButtonStyled = styled.button`
    position: relative;
    font-size: ${({
        theme: {
            fontSizes: { extraExtraLarge },
        },
    }) => extraExtraLarge};
    padding: ${({
        theme: {
            spacing: { small, large },
        },
    }) => `${small} ${large}`};
    font-weight: 800;
    transition: color 0.3s ease, background-color 0.3s ease;
    background-color: ${({
        color,
        theme: {
            colors,
            colors: { pink },
        },
    }) => (color ? colors[`${color}`] : pink)};
    border-width: 0.12rem;
    border-style: solid;
    border-radius: 0.5rem;
    border-color: ${({
        color,
        theme: {
            colors,
            colors: { pink },
        },
    }) => (color ? colors[`${color}`] : pink)};
    color: ${({
        color,
        theme: {
            colors: { white, darkBlue },
        },
    }) => {
        switch (color) {
            case 'white':
                return darkBlue;
            case 'lightBlue':
                return darkBlue;
            case 'blue':
                return white;
            case 'pink':
                return white;
            case 'darkGrey':
                return white;
            case undefined:
                return white;
            default:
                return white;
        }
    }};

    width: 100%;

    @media screen and (min-width: ${({
            theme: {
                breakpoints: { small },
            },
        }) => small}) {
        width: auto;
    }

    &:hover,
    &:focus {
        background-color: ${({
            color,
            theme: {
                colors: { white },
            },
        }) => {
            switch (color) {
                case 'white':
                    return `rgba(0, 0, 0, 0.4)`;
                case 'lightBlue':
                    return `rgba(0, 0, 0, 0.4)`;
                case 'blue':
                    return white;
                case 'pink':
                    return white;
                case 'darkGrey':
                    return white;
                case undefined:
                    return white;
                default:
                    return white;
            }
        }};
        color: ${({
            color,
            theme: {
                colors,
                colors: { pink },
            },
        }) => (color ? colors[`${color}`] : pink)};
    }

    &::after {
        content: '';
        display: ${({ isShiny }: { isShiny: boolean }) =>
            isShiny ? `block` : `none`};
        position: absolute;
        top: 0.05%;
        left: 0.05%;
        width: calc(100% - 0.1%);
        height: 50%;
        border-radius: 0.5rem;
        background: linear-gradient(
            rgba(255, 255, 255, 0.5),
            rgba(255, 255, 255, 0.2)
        );
    }
`;
