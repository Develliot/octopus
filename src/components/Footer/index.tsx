import {
    ColorBlock,
    ContentWrapper,
    VerticalSpacer,
} from 'src/components/Layout';
import { Paragraph } from 'src/components/Typography';

export const Footer = () => (
    <footer>
        <ColorBlock padding={false}>
            <ContentWrapper>
                <VerticalSpacer size='small' />
                <Paragraph fontSize='tiny' style={{ fontWeight: 200 }}>
                    Octopus Energy Ltd is a company registered in England and
                    Wales.
                </Paragraph>
                <Paragraph fontSize='tiny' style={{ fontWeight: 200 }}>
                    Registered number: 09263424. Registered office: 33 Holborn,
                    London, EC1N 2HT. Trading office: 20-24 Broadwick Street,
                    London, W1F 8HT.
                </Paragraph>
                <VerticalSpacer size='small' />
            </ContentWrapper>
        </ColorBlock>
    </footer>
);

export default Footer;
