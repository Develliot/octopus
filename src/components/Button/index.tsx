import React, { FunctionComponent } from 'react';
import { ColorType } from 'src/theme';

import { ButtonStyled } from './styles';

type Props = {
    color?: ColorType;
    isShiny?: boolean;
    onClick?: () => void;
};

export const Button: FunctionComponent<Props> = ({
    color = 'blue',
    isShiny = false,
    onClick = () => null,
    children,
}) => {
    return (
        <ButtonStyled onClick={onClick} color={color} isShiny={isShiny}>
            {children}
        </ButtonStyled>
    );
};

export default Button;
