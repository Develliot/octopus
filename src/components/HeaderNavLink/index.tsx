import React from 'react';
import { Paragraph } from 'src/components/Typography';
import { RowCenter } from 'src/components/Layout';
import { NavLinkStyled, HoverIndicator, HeaderLinkItem } from './styles';
import Link from 'next/link';

export const HeaderNavLink = ({
    to,
    title,
    selected,
    onClick,
}: {
    to: string;
    title: string;
    selected: boolean;
    onClick?: () => void;
}) => {
    const handleClick = !!onClick ? onClick : () => {};

    return (
        <HeaderLinkItem>
            <Link href={to}>
                <NavLinkStyled
                    title={title}
                    className={selected ? 'selected' : ''}
                    onClick={handleClick}
                >
                    <Paragraph color='white'>{title}</Paragraph>
                    <RowCenter>
                        <HoverIndicator selected={selected} />
                    </RowCenter>
                </NavLinkStyled>
            </Link>
        </HeaderLinkItem>
    );
};
