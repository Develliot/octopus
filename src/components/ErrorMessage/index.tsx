import React, { FunctionComponent } from 'react';
import Link from 'next/link';

import urls from 'src/urls';

import { Button } from 'src/components/Button';
import { H2, Paragraph } from 'src/components/Typography';
import { VerticalSpacer } from 'src/components/Layout';

import { ErrorWrapper } from './styles';

type Props = {
    errorMessage: string;
    retry?: () => void;
};

export const ErrorMessage: FunctionComponent<Props> = ({
    errorMessage,
    retry,
}) => {
    return (
        <ErrorWrapper>
            <H2 color='darkBlue'>Something broke!</H2>
            <VerticalSpacer size='large' />
            <Paragraph color='darkBlue'>{errorMessage}</Paragraph>
            <VerticalSpacer size='large' />
            {retry ? (
                <>
                    <Button color='pink' onClick={retry} isShiny={true}>
                        Retry
                    </Button>
                    <VerticalSpacer size='medium' />
                </>
            ) : null}
            <Link href={urls.home}>
                <a title='home'>
                    <Button isShiny={true}>Home</Button>
                </a>
            </Link>
        </ErrorWrapper>
    );
};

export default ErrorMessage;
