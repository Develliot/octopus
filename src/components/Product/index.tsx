import { useState, useEffect, memo, useRef } from 'react';

import { useBasketContext } from 'src/contexts/BasketContext';
import { BasketReducer } from 'src/reducers/BasketReducer';

import { Button } from 'src/components/Button';
import { H1, H2, Paragraph } from 'src/components/Typography';
import {
    VerticalSpacer,
    RowCenter,
    ColorBlock,
    ContentWrapper,
    RowStretch,
} from 'src/components/Layout';
import Price from 'src/components/Price';
import QuantityCounter from 'src/components/QuantityCounter';
import ShowMore from 'src/components/ShowMore';

import {
    CoverColor,
    ImageWrapper,
    ImageContainer,
    SpecTable,
    SpecRow,
} from './styles';

import { ProductType } from 'src/types';

type Props = {
    product: ProductType;
};

const MIN_QUANTITY = 1;

export const Product = ({
    product,
    product: {
        imageUrl,
        title,
        wattage,
        packetSize,
        price,
        description,
        specifications,
    },
}: Props) => {
    const [count, setCount] = useState(MIN_QUANTITY);
    const [hasBeenAddedToBasket, setHasBeenAddedToBasket] = useState(false);
    const [state, setState] = useBasketContext();
    const ref = useRef();

    // hide added to basket message after a few seconds
    useEffect(() => {
        if (hasBeenAddedToBasket) {
            setTimeout(() => {
                // we need to check ref incase component unmounts during timeout
                if (ref) {
                    setHasBeenAddedToBasket(false);
                }
            }, 3000);
        }
    }, [ref, hasBeenAddedToBasket, setHasBeenAddedToBasket]);

    const handleQuantityChanged = (value: number) => {
        setCount(value);
    };

    const handleAddToCart = () => {
        const newState = BasketReducer(state, {
            type: 'addProducts',
            payload: { product: product, value: count },
        });
        setState(newState);
        setHasBeenAddedToBasket(true);
    };

    return (
        <>
            <RowCenter>
                <CoverColor>
                    <RowCenter>
                        <ImageWrapper>
                            <ImageContainer src={imageUrl} alt={title} />
                        </ImageWrapper>
                    </RowCenter>
                </CoverColor>
            </RowCenter>
            <ColorBlock padding={false}>
                <ContentWrapper>
                    <VerticalSpacer size='large' />
                    <H1>{title}</H1>
                    <VerticalSpacer size='tiny' />
                    <Paragraph color='blueGrey'>
                        {`${wattage}W // Packet of ${packetSize}`}
                    </Paragraph>
                </ContentWrapper>
                <VerticalSpacer size='small' />
            </ColorBlock>
            <ColorBlock color='darkBlue' padding={false}>
                <ContentWrapper>
                    <VerticalSpacer size='small' />
                    <RowStretch>
                        <div>
                            <VerticalSpacer size='medium' />
                            <Price price={price} />
                        </div>
                        <QuantityCounter
                            minValue={MIN_QUANTITY}
                            onChange={handleQuantityChanged}
                        />
                    </RowStretch>
                    <VerticalSpacer size='small' />
                    <Button
                        isShiny={true}
                        color='pink'
                        onClick={handleAddToCart}
                    >
                        Add to cart
                    </Button>
                    <VerticalSpacer size='large'>
                        {hasBeenAddedToBasket && (
                            <Paragraph color='pink'>
                                ...added to basket
                            </Paragraph>
                        )}
                    </VerticalSpacer>
                </ContentWrapper>
            </ColorBlock>
            <ColorBlock>
                <ContentWrapper>
                    <H2>Description</H2>
                    <VerticalSpacer size='medium' />
                    <ShowMore color='blue' closeHeight='4em'>
                        <Paragraph ellipsis={true}>{description}</Paragraph>
                    </ShowMore>
                </ContentWrapper>
            </ColorBlock>

            <ColorBlock color='darkBlue'>
                <ContentWrapper>
                    <H2>Specifications</H2>
                    <VerticalSpacer size='medium' />
                    <ShowMore color='darkBlue' closeHeight='4em'>
                        <SpecTable>
                            <tbody>
                                <SpecRow>
                                    <td>Brand</td>
                                    <td>{specifications.brand}</td>
                                </SpecRow>
                                <SpecRow>
                                    <td>Item weight</td>
                                    <td>{specifications.weight}</td>
                                </SpecRow>
                                <SpecRow>
                                    <td>Dimensions</td>
                                    <td>{specifications.dimensions}</td>
                                </SpecRow>
                                <SpecRow>
                                    <td>Item model number</td>
                                    <td>{specifications.modelNumber}</td>
                                </SpecRow>
                                <SpecRow>
                                    <td>Colour</td>
                                    <td>{specifications.colour}</td>
                                </SpecRow>
                            </tbody>
                        </SpecTable>
                    </ShowMore>
                </ContentWrapper>
            </ColorBlock>
        </>
    );
};

export default memo(Product);
