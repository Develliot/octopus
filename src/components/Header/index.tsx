import React, { useState } from 'react';
import Link from 'next/link';
import { useRouter } from 'next/router';

import { useScrollPosition } from 'src/hooks/useScrollPosition';
import urls from 'src/urls';

import { RowStretch } from 'src/components/Layout';
import { HeaderNavLink } from 'src/components/HeaderNavLink';
import { MenuButton } from './MenuButton';

import {
    HeaderWrapper,
    HeaderInner,
    Links,
    LogoSVGStyled,
    ResponsiveSpacer,
} from './styles';

export const Header = ({
    visible,
    currentRoute,
}: {
    visible: boolean;
    currentRoute: string;
}) => {
    const [isMenuOpen, setIsMenuOpen] = useState(false);

    const isSelected = (url: string): boolean => url === currentRoute;
    const handleOnClick = () => {
        setIsMenuOpen(!isMenuOpen);
    };

    const closeMenu = () => {
        setIsMenuOpen(false);
    };

    return (
        <HeaderWrapper visible={visible}>
            <HeaderInner>
                <RowStretch>
                    <Link href={urls.home}>
                        <a title='home' tabIndex={0}>
                            <LogoSVGStyled />
                        </a>
                    </Link>
                    <MenuButton onClick={handleOnClick} />
                </RowStretch>
                <ResponsiveSpacer menuOpen={isMenuOpen} />
                <nav>
                    <Links menuOpen={isMenuOpen}>
                        <HeaderNavLink
                            to={urls.home}
                            title='Shop'
                            selected={isSelected(urls.home)}
                            onClick={closeMenu}
                        />
                        <HeaderNavLink
                            to={urls.basket}
                            title='Basket'
                            selected={isSelected(urls.basket)}
                            onClick={closeMenu}
                        />
                        <HeaderNavLink
                            to={urls.designSystem}
                            title='Design System'
                            selected={isSelected(urls.designSystem)}
                            onClick={closeMenu}
                        />
                    </Links>
                </nav>
            </HeaderInner>
        </HeaderWrapper>
    );
};

export const HideOnScroll = () => {
    const [hideOnScroll, setHideOnScroll] = useState(true);
    const router = useRouter();

    useScrollPosition(
        ({
            prevPos,
            currPos,
        }: {
            prevPos: { x: number; y: number };
            currPos: { x: number; y: number };
        }) => {
            const isShow = currPos.y >= prevPos.y;
            if (isShow !== hideOnScroll) setHideOnScroll(isShow);
        },
        [hideOnScroll],
        null,
        false,
        300
    );

    return <Header visible={hideOnScroll} currentRoute={router.route} />;
};

export default HideOnScroll;
