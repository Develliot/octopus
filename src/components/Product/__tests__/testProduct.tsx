import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';

import { Product } from '../';
import { Button } from 'src/components/Button';
import { Paragraph } from 'src/components/Typography';

import { ProductType } from 'src/types';
import * as BasketContext from 'src/contexts/BasketContext';

const mockProduct: ProductType = {
    id: 1,
    title: 'Energy saving light bulb',
    wattage: 25,
    packetSize: 4,
    price: 12.99,
    description:
        'Available in 7 watts, 9 watts, 11 watts Spiral Light bulb in B22, bulb switches instantly, no wait around with a warm start and flicker free features. Light up your life with this swirly wonder!',
    imageUrl: 'https://picsum.photos/500/500',
    specifications: {
        brand: 'Philips',
        weight: '77g',
        dimensions: '12.6x6.2x6.2cm',
        modelNumber: 'E27 ES',
        colour: 'Cool daylight',
    },
};

describe('Product', () => {
    it('renders correctly with product', () => {
        const wrapper: ShallowWrapper = shallow(
            <Product product={mockProduct} />
        );
        expect(wrapper).toMatchSnapshot();
    });

    it('displays when an item has been added to cart', () => {
        const wrapper: ShallowWrapper = shallow(
            <Product product={mockProduct} />
        );
        wrapper.find(Button).simulate('click');
        expect(
            wrapper
                .find(Paragraph)
                .at(1)
                .text()
        ).toContain('...added to basket');
    });

    it('sends data to the store when add to basked its clicked', () => {
        const contextValues: BasketContext.BasketContextStateType = {
            basket: [{ product: mockProduct, quantity: 3 }],
        };

        const dispatch = jest.fn();

        jest.spyOn(BasketContext, 'useBasketContext').mockImplementation(() => {
            return [contextValues, dispatch];
        });

        const wrapper: ShallowWrapper = shallow(
            <Product product={mockProduct} />
        );
        wrapper.find(Button).simulate('click');
        expect(dispatch).toHaveBeenCalledWith({
            basket: [{ product: mockProduct, quantity: 4 }],
        });
    });
});
