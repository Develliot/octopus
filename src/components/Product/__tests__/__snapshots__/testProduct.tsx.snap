// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product renders correctly with product 1`] = `
<Fragment>
  <Layout__RowCenter>
    <styles__CoverColor>
      <Layout__RowCenter>
        <styles__ImageWrapper>
          <styles__ImageContainer
            alt="Energy saving light bulb"
            src="https://picsum.photos/500/500"
          />
        </styles__ImageWrapper>
      </Layout__RowCenter>
    </styles__CoverColor>
  </Layout__RowCenter>
  <ColorBlock
    padding={false}
  >
    <Layout__ContentWrapper>
      <Layout__VerticalSpacer
        size="large"
      />
      <Typography__H1>
        Energy saving light bulb
      </Typography__H1>
      <Layout__VerticalSpacer
        size="tiny"
      />
      <Typography__Paragraph
        color="blueGrey"
      >
        25W // Packet of 4
      </Typography__Paragraph>
    </Layout__ContentWrapper>
    <Layout__VerticalSpacer
      size="small"
    />
  </ColorBlock>
  <ColorBlock
    color="darkBlue"
    padding={false}
  >
    <Layout__ContentWrapper>
      <Layout__VerticalSpacer
        size="small"
      />
      <Layout__RowStretch>
        <div>
          <Layout__VerticalSpacer
            size="medium"
          />
          <Memo(Price)
            price={12.99}
          />
        </div>
        <QuantityCounter
          minValue={1}
          onChange={[Function]}
        />
      </Layout__RowStretch>
      <Layout__VerticalSpacer
        size="small"
      />
      <Button
        color="pink"
        isShiny={true}
        onClick={[Function]}
      >
        Add to cart
      </Button>
      <Layout__VerticalSpacer
        size="large"
      />
    </Layout__ContentWrapper>
  </ColorBlock>
  <ColorBlock>
    <Layout__ContentWrapper>
      <Typography__H2>
        Description
      </Typography__H2>
      <Layout__VerticalSpacer
        size="medium"
      />
      <Typography__Paragraph>
        Available in 7 watts, 9 watts, 11 watts Spiral Light bulb in B22, bulb switches instantly, no wait around with a warm start and flicker free features. Light up your life with this swirly wonder!
      </Typography__Paragraph>
    </Layout__ContentWrapper>
  </ColorBlock>
  <ColorBlock
    color="darkBlue"
  >
    <Layout__ContentWrapper>
      <Typography__H2>
        Specifications
      </Typography__H2>
      <Layout__VerticalSpacer
        size="medium"
      />
      <styles__SpecTable>
        <tbody>
          <styles__SpecRow>
            <td>
              Brand
            </td>
            <td>
              Philips
            </td>
          </styles__SpecRow>
          <styles__SpecRow>
            <td>
              Item weight
            </td>
            <td>
              77g
            </td>
          </styles__SpecRow>
          <styles__SpecRow>
            <td>
              Dimensions
            </td>
            <td>
              12.6x6.2x6.2cm
            </td>
          </styles__SpecRow>
          <styles__SpecRow>
            <td>
              Item model number
            </td>
            <td>
              E27 ES
            </td>
          </styles__SpecRow>
          <styles__SpecRow>
            <td>
              Colour
            </td>
            <td>
              Cool daylight
            </td>
          </styles__SpecRow>
        </tbody>
      </styles__SpecTable>
    </Layout__ContentWrapper>
  </ColorBlock>
</Fragment>
`;
